wp-includes/mysql/class-mysql-lexer.php

```diff
@@ -0,0 +1,2752 @@
+<?php
+
+require_once __DIR__ . '/class-mysql-token.php';
+
+/**
+ * This lexer is based on the MySQL Workbench lexer grammar.
+ * See:
+ *   https://github.com/mysql/mysql-workbench/blob/8.0.38/library/parsers/grammars/MySQLLexer.g4
+ *   https://github.com/mysql/mysql-workbench/blob/8.0.38/library/parsers/grammars/predefined.tokens
+ *   https://github.com/mysql/mysql-workbench/blob/8.0.38/library/parsers/mysql/MySQLBaseLexer.cpp
+ */
+class MySQL_Lexer {
+	// Token channels
+	const CHANNEL_DEFAULT = 0;
+	const CHANNEL_HIDDEN  = 99;
+
+	// SQL modes
+	const SQL_MODE_NO_MODE              = 0;
+	const SQL_MODE_ANSI_QUOTES          = 1 << 0;
+	const SQL_MODE_HIGH_NOT_PRECEDENCE  = 1 << 1;
+	const SQL_MODE_PIPES_AS_CONCAT      = 1 << 2;
+	const SQL_MODE_IGNORE_SPACE         = 1 << 3;
+	const SQL_MODE_NO_BACKSLASH_ESCAPES = 1 << 4;
```

Comment:
Nitpick: what do you think about using number literals here, like `16`? I noticed myself running the `1 << 4` computation in my head as I scrolled and for a second I wasn't sure if it's "1 shifted 4 bits to the left" or "4 shifted 1 bit to the left".